### PR TITLE
convos: init at 4.22

### PR DESCRIPTION
**Pull request by stigtsp**
This PR add the web based IRC client `convos`, including nixos module and tests.

Several perlPackage dependencies are added, and some are updated.

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [X] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by stigok — APPROVED** [review]

**Review by fnords — APPROVED** [review]

**Comment by stigtsp** [comment]
Updating convos to 4.18

File: nixos/modules/services/web-apps/convos.nix
Comment (jhthorsen) [inline nixos/modules/services/web-apps/convos.nix:44]: `X-Forwarded-Proto` is not required. `X-Request-Base` is more important.

File: pkgs/applications/networking/irc/convos/default.nix
Comment (jhthorsen) [inline pkgs/applications/networking/irc/convos/default.nix:7]: Just out of curiosity: Is there a way to make this dynamic? Sorry if this is off topic and something I should look up or ask in a different channel.

File: pkgs/top-level/perl-packages.nix
Comment (jhthorsen) [inline pkgs/top-level/perl-packages.nix:12472]: Mojo::IRC is no longer required by Convos.

**Review by jhthorsen — COMMENTED** [review]
This is awesome 👍 So glad you created this PR @stigtsp!

File: nixos/modules/services/web-apps/convos.nix
Comment (stigtsp) [reply nixos/modules/services/web-apps/convos.nix]: Changed the texts slightly

File: pkgs/applications/networking/irc/convos/default.nix
Comment (stigtsp) [reply pkgs/applications/networking/irc/convos/default.nix]: Updated

File: pkgs/applications/networking/irc/convos/default.nix
Comment (stigtsp) [reply pkgs/applications/networking/irc/convos/default.nix:7]: Thx, replaced with the text on the website.

File: pkgs/top-level/perl-packages.nix
Comment (stigtsp) [reply pkgs/top-level/perl-packages.nix]: Thx for spotting this. I've removed Mojo::IRC from this PR.

File: nixos/modules/services/web-apps/convos.nix
Comment (stigtsp) [reply nixos/modules/services/web-apps/convos.nix]: Thx, updated.

File: pkgs/applications/networking/irc/convos/default.nix
Comment (stigtsp) [reply pkgs/applications/networking/irc/convos/default.nix]: Sure, bumped to 4.19

File: nixos/modules/services/web-apps/convos.nix
Comment (aanderse) [inline nixos/modules/services/web-apps/convos.nix:48]: Does this actually need a permanent user account, or would `DynamicUser` suffice?

File: nixos/modules/services/web-apps/convos.nix
Comment (aanderse) [inline nixos/modules/services/web-apps/convos.nix:18]: `types.port` exists.

File: nixos/modules/services/web-apps/convos.nix
Comment (aanderse) [inline nixos/modules/services/web-apps/convos.nix:11]: It doesn't seem like there would ever be a practical case to override the default here, so I would recommend dropping it.

File: nixos/modules/services/web-apps/convos.nix
Comment (aanderse) [inline nixos/modules/services/web-apps/convos.nix:54]: You could move this chunk of configuration from `serviceConfig` into `systemd.services.convos` and write it the following way instead:

```suggestion
        environment = {
          CONVOS_HOME = "%S/convos";
          CONVOS_REVERSE_PROXY = if cfg.reverseProxy then "1" else "0"; # I'm guessing simply putting cfg.reverseProxy; would even work here...
          MOJO_LISTEN = "http://${toString cfg.listenAddress}:${toString cfg.listenPort}";
        };
```

**Review by aanderse — CHANGES_REQUESTED** [review]
I have reviewed the module. Maybe @volth wants to give a quick glance over the packages for final approval there :man_shrugging: 

Looking good though :+1:

File: nixos/modules/services/web-apps/convos.nix
Comment (stigtsp) [reply nixos/modules/services/web-apps/convos.nix:48]: Good point. I think it should work, and makes some of the other hardening options redundant. I'll need to test it a bit and update the commit.

File: nixos/modules/services/web-apps/convos.nix
Comment (stigtsp) [reply nixos/modules/services/web-apps/convos.nix:54]: Thanks, that seems a bit nicer! Will do

File: nixos/modules/services/web-apps/convos.nix
Comment (stigtsp) [reply nixos/modules/services/web-apps/convos.nix]: Works well - I've tweaked the Protect* and Restrict* options a bit.

**Comment by stigtsp** [comment]
Thx for reviewing :+1:

File: nixos/modules/services/web-apps/convos.nix
Comment (aanderse) [inline nixos/modules/services/web-apps/convos.nix:53]: @stigtsp at least a few of these options are already implied with `DynamicUser`. A quick search for `DynamicUser` in the `systemd` manual will cut a few lines here.

Comment (talyz) [reply nixos/modules/services/web-apps/convos.nix:53]: Yes, `ProtectSystem` and `ProtectHome` are implied and can be removed.

Comment (stigtsp) [reply nixos/modules/services/web-apps/convos.nix:52]: Thanks @talyz @aanderse :)  I've removed `ProtectSystem=strict`.

Seems like `ProtectHome=read-only` is implied by `DynamicUser=true` according to the documentation, so keeping `ProtectHome=true`.

Added `SystemCallFilter`, `SystemCallArchitectures`, `CapabilityBoundingSet`, and some more flags highlighted by `systemd-analyze security`.

```
→ Overall exposure level for convos.service: 1.3 OK 🙂
```

Does this look ok?

Comment (talyz) [reply nixos/modules/services/web-apps/convos.nix:52]: Ah, yes, you're right :) Looks good to me, assuming all features you want still work as expected ;)

**Comment by talyz** [comment]
@GrahamcOfBorg test convos

**Comment by stigtsp** [comment]
Waiting for a new release from upstream that contains some important fixes.

**Comment by talyz** [comment]
@stigtsp Can you add the test to `passthru.tests`? That way the test is automatically run for PRs and easy to refer to. See `tests` here: https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes

**Comment by stigtsp** [comment]
This PR includes updates to `perlPackages.MojoliciousPluginOpenAPI`, `perlPackages.JSONValidator` and `perlPackages.Mojolicious` which are also included in https://github.com/NixOS/nixpkgs/pull/91184 I suggest  merging this PR after the `cpan2nix` updates have been merged.

**Comment by volth** [comment]
> I suggest merging this PR after the `cpan2nix` updates have been merged.

Those mass updates are typically stuck for 2-3 weeks (no one wants to review them), so probably there is no point to wait.
Also that PR is against `staging` which merged to `master` every few weeks.

**Comment by stigtsp** [comment]
> Those mass updates are typically stuck for 2-3 weeks (no one wants to review them), so probably there is no point to wait.
> Also that PR is against `staging` which merged to `master` every few weeks.

Ok - updated Mojolicious in this PR to 8.55, so it should be ready iIhope :-) Does the perlPackages update look ok?

Result of `nixpkgs-review pr 88940` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>42 packages built:</summary>
<br>- abcde
<br>- convos
<br>- perl528Packages.IRCUtils
<br>- perl528Packages.JSONValidator
<br>- perl528Packages.LinkEmbedder
<br>- perl528Packages.MojoIOLoopForkCall
<br>- perl528Packages.MojoJWT
<br>- perl528Packages.MojoPg
<br>- perl528Packages.MojoRedis
<br>- perl528Packages.MojoSQLite
<br>- perl528Packages.Mojolicious
<br>- perl528Packages.MojoliciousPluginMail
<br>- perl528Packages.MojoliciousPluginOpenAPI
<br>- perl528Packages.MojoliciousPluginStatus
<br>- perl528Packages.MojoliciousPluginTextExceptions
<br>- perl528Packages.MojoliciousPluginWebpack
<br>- perl528Packages.Mojomysql
<br>- perl528Packages.MusicBrainz
<br>- perl528Packages.OpenAPIClient
<br>- perl528Packages.ParseIRC
<br>- perl528Packages.TimePiece
<br>- perl528Packages.UnicodeUTF8
<br>- perl530Packages.IRCUtils
<br>- perl530Packages.JSONValidator
<br>- perl530Packages.LinkEmbedder
<br>- perl530Packages.MojoIOLoopForkCall
<br>- perl530Packages.MojoJWT
<br>- perl530Packages.MojoPg
<br>- perl530Packages.MojoRedis
<br>- perl530Packages.MojoSQLite
<br>- perl530Packages.Mojolicious
<br>- perl530Packages.MojoliciousPluginMail
<br>- perl530Packages.MojoliciousPluginOpenAPI
<br>- perl530Packages.MojoliciousPluginStatus
<br>- perl530Packages.MojoliciousPluginTextExceptions
<br>- perl530Packages.MojoliciousPluginWebpack
<br>- perl530Packages.Mojomysql
<br>- perl530Packages.MusicBrainz
<br>- perl530Packages.OpenAPIClient
<br>- perl530Packages.ParseIRC
<br>- perl530Packages.TimePiece
<br>- perl530Packages.UnicodeUTF8
</details>

**Review by aanderse — APPROVED** [review]
Yeah this looks good to me from a quick overview. @talyz I think we should merge... agree?

**Review by talyz — APPROVED** [review]
@aanderse Yep, looks good to me :+1:

**Comment by talyz** [comment]
@stigtsp Thanks for doing this! Great work! :tada:

**Comment by stigtsp** [comment]
Thx for the reviews and advice, everyone :)

**Comment by FRidh** [comment]
Fixed eval in cd8e099ffeee27d229371f4e8cba1a8b155e2a48.

**Pull request merged**
